rfc-113-expanding-draft-access-for-unauthenticated-users.md

```diff
@@ -0,0 +1,212 @@
+  authenticated user has access to a resource if an auth-bypass token is
+  present. This introduces potential confusion when this token is provided by
+  a session cookie and could prevent an authenticated user viewing other parts
+  of draft GOV.UK. To counter this both apps will be changed to consider both
+  authentication and auth-bypass token when considering access to resources.
+- Content Store will be updated to allow an auth-bypass id to be present in the
+  links of a content item, this will be checked against a user provided
+  auth-bypass_id for content items that are not access limited.
+- Publishing API will be changed to support the sharing of auth-bypass ids for an
+  edition through link expansion rules.
+- Publishing API and Content Store will have their access limit code decoupled
+  from auth-bypass ids. This will be done to remove a frequent source of
+  confusion that conflates the separate processes of access limiting and
+  auth-bypass.
+- Collections Publisher will be changed to include a flag in auth-bypass tokens
+  that provide draft asset access. To reflect the increased access of a token
+  they will be given an expiry time (anticipated to be 1 month). Collections
+  Publisher will also be updated to include origin information in the token
+  (such as creator id and content id) to allow a token to be audited. The
+  [auth-bypass token documentation][auth-bypass-token-docs] will be updated to
+  reflect this as the best practice for future auth-bypass tokens
+
+### Rejected alternative approaches
+
+There were a number of alternative ideas which were considered to provide fact
+checking for step-by-steps and were ultimately rejected. This is a quick
+compilation of them and the reasons why they were rejected.
+
+#### Providing a link to preview the embedded part of a step-by-step distinct from the content
+
+This would involve creating a page which shows just the step-by-step
+navigation component that is embedded on linked pages. This was deemed
+insufficient as fact checkers are often comparing the step-by-step navigation
+with the context of the specific guidance pages.
+
+#### Providing a means to embed the step-by-step component into live content
+
+This would involve a means of injecting the step-by-step navigation component
+into live GOV.UK pages via JavaScript. This would prevent a fact checker of a
+step-by-step from viewing any draft content except the step-by-step
+navigation itself. This was quickly rejected as it was considered to be a
+common scenario for the pages that are part of a step-by-step
+to also be drafts.
+
+#### Exclude asset access from fact-checking
+
+Assets are the area of access that provide the most complication since they
+are distinct from the Publishing API link system. An option was to exclude
+them from the access checks and have them return forbidden errors if a
+step-by-step page used them - this would typically be most noticeable on an
+embedded image.
```

Comment:
Probably another rejection worth documenting - couldn't we write a task which goes through all content items, looks for asset URLs, retrieves the associated UIDs of the content items and saves those as links in publishing API?

That would then free up possibilities for providing tokens for assets via link expansion.

Reply:
> Probably another rejection worth documenting - couldn't we write a task which goes through all content items, looks for asset URLs, retrieves the associated UIDs of the content items and saves those as links in publishing API?
> 
> That would then free up possibilities for providing tokens for assets via link expansion.

We don't have a concept of links to assets unfortunately - that'd have resolved the need for the draft_asset_manager_access grant. I like to imagine that at some point in the future we actually merge Asset Manager and Publishing API together since they're very similar concepts for providing draft and live resources on GOV.UK and their subtle differences are points of confusion.

With how things currently stand though app wise, the way we could give each asset a token would be to do the following at the point of sending a draft from Publishing API to Content Store:

- Determine whether any links of the edition provide auth_bypass_ids
- Parse all fields that could contain URLs in the Publishing API (this might be as grim as turning details JSON into a  string and pattern matching for URLs) and pick out any that are to assets
- Query asset manager for each of these to see if they are draft
- For all the draft ones send an update request to the asset to set the auth_bypass_id

This is not an unworkable approach but it is undesirable as it'll add a new Asset Manager dependency on Publishing API and add extra time and complexity to the downstream process.